src/Chapter3.hs

```diff
@@ -480,6 +517,41 @@ After defining the city, implement the following functions:
    and at least 10 living people inside
 -}
 
+data OptionalCastle = NoCastle | Castle String deriving (Show, Eq)
+data OptionalWall = NoWall | Wall deriving (Show, Eq)
+data RestHouse = Church | Library deriving (Show, Eq)
+data House = MkHouse 
+  { houseLiving :: Int
+  } deriving (Show, Eq)
+
+data MagicalCity = MkCity
+  { cityCastle :: OptionalCastle
+  , cityWall :: OptionalWall
```

Comment:
Current data types are not typesafe enough. By the definition of your `MagicalCity` it is possible to have `cityCastle = NoCastle` and `cityWall = Wall`. It would be more safe if you had a `Wall` indicator inside the `Castle` type,  so it would be impossible to create walls without a castle